In [ ]:
from google.colab import drive
drive.mount("/content/gdrive")

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

pd.set_option("display.max_columns",None)
pd.set_option("display.max_rows",None)
df_a=pd.read_csv("/content/gdrive/MyDrive/Colab Notebooks/application_data.csv")
df_p=pd.read_csv("/content/gdrive/MyDrive/Colab Notebooks/previous_application.csv")

In [ ]:
#knowing the data
print("top 5 rows of data:\n\n",df_a.head(5))

top 5 rows of data:

    SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
0      100002       1         Cash loans           M            N   
1      100003       0         Cash loans           F            N   
2      100004       0    Revolving loans           M            Y   
3      100006       0         Cash loans           F            N   
4      100007       0         Cash loans           M            N   

  FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0               Y             0          202500.0    406597.5      24700.5   
1               N             0          270000.0   1293502.5      35698.5   
2               Y             0           67500.0    135000.0       6750.0   
3               Y             0          135000.0    312682.5      29686.5   
4               Y             0          121500.0    513000.0      21865.5   

   AMT_GOODS_PRICE NAME_TYPE_SUITE NAME_INCOME_TYPE  \
0         351000.0   Unaccompanied     

In [ ]:
print("Getting information about column\n\n", df_a.dtypes)

Getting information about column

 SK_ID_CURR                        int64
TARGET                            int64
NAME_CONTRACT_TYPE               object
CODE_GENDER                      object
FLAG_OWN_CAR                     object
FLAG_OWN_REALTY                  object
CNT_CHILDREN                      int64
AMT_INCOME_TOTAL                float64
AMT_CREDIT                      float64
AMT_ANNUITY                     float64
AMT_GOODS_PRICE                 float64
NAME_TYPE_SUITE                  object
NAME_INCOME_TYPE                 object
NAME_EDUCATION_TYPE              object
NAME_FAMILY_STATUS               object
NAME_HOUSING_TYPE                object
REGION_POPULATION_RELATIVE      float64
DAYS_BIRTH                        int64
DAYS_EMPLOYED                     int64
DAYS_REGISTRATION               float64
DAYS_ID_PUBLISH                   int64
OWN_CAR_AGE                     float64
FLAG_MOBIL                        int64
FLAG_EMP_PHONE                    int64
FLAG_

In [ ]:
print("Getting name of all columns: \n\n",df_a.columns)

Getting name of all columns: 

 Index(['SK_ID_CURR', 'TARGET', 'NAME_CONTRACT_TYPE', 'CODE_GENDER',
       'FLAG_OWN_CAR', 'FLAG_OWN_REALTY', 'CNT_CHILDREN', 'AMT_INCOME_TOTAL',
       'AMT_CREDIT', 'AMT_ANNUITY',
       ...
       'FLAG_DOCUMENT_18', 'FLAG_DOCUMENT_19', 'FLAG_DOCUMENT_20',
       'FLAG_DOCUMENT_21', 'AMT_REQ_CREDIT_BUREAU_HOUR',
       'AMT_REQ_CREDIT_BUREAU_DAY', 'AMT_REQ_CREDIT_BUREAU_WEEK',
       'AMT_REQ_CREDIT_BUREAU_MON', 'AMT_REQ_CREDIT_BUREAU_QRT',
       'AMT_REQ_CREDIT_BUREAU_YEAR'],
      dtype='object', length=122)


In [ ]:
print(f"no. of row:{df_a.shape[0]}\n no. of columns: {df_a.shape[1]}")

no. of row:307511
 no. of columns: 122


In [ ]:
#checking for nan values
(df_a.isna().sum()/len(df_a)*100).sort_values(axis=0)   

SK_ID_CURR                       0.000000
HOUR_APPR_PROCESS_START          0.000000
REG_REGION_NOT_WORK_REGION       0.000000
LIVE_REGION_NOT_WORK_REGION      0.000000
REG_CITY_NOT_LIVE_CITY           0.000000
REG_CITY_NOT_WORK_CITY           0.000000
LIVE_CITY_NOT_WORK_CITY          0.000000
ORGANIZATION_TYPE                0.000000
FLAG_DOCUMENT_21                 0.000000
FLAG_DOCUMENT_20                 0.000000
FLAG_DOCUMENT_19                 0.000000
FLAG_DOCUMENT_18                 0.000000
FLAG_DOCUMENT_17                 0.000000
FLAG_DOCUMENT_16                 0.000000
FLAG_DOCUMENT_15                 0.000000
FLAG_DOCUMENT_14                 0.000000
FLAG_DOCUMENT_13                 0.000000
FLAG_DOCUMENT_12                 0.000000
FLAG_DOCUMENT_11                 0.000000
FLAG_DOCUMENT_10                 0.000000
FLAG_DOCUMENT_9                  0.000000
FLAG_DOCUMENT_8                  0.000000
FLAG_DOCUMENT_7                  0.000000
FLAG_DOCUMENT_6                  0

In [ ]:
print("columns with more than 50% of Nan values \n\n")
na_val=((df_a.isna().sum()/len(df_a))*100).sort_values()>50
print("count of columns:",na_val[na_val.values==True].count(),"\n\n ")
print(na_val[na_val.values==True])

columns with more than 50% of Nan values 


count of columns: 41 

 
HOUSETYPE_MODE              True
LIVINGAREA_AVG              True
LIVINGAREA_MODE             True
LIVINGAREA_MEDI             True
ENTRANCES_AVG               True
ENTRANCES_MODE              True
ENTRANCES_MEDI              True
APARTMENTS_MEDI             True
APARTMENTS_AVG              True
APARTMENTS_MODE             True
WALLSMATERIAL_MODE          True
ELEVATORS_MEDI              True
ELEVATORS_AVG               True
ELEVATORS_MODE              True
NONLIVINGAREA_MODE          True
NONLIVINGAREA_AVG           True
NONLIVINGAREA_MEDI          True
EXT_SOURCE_1                True
BASEMENTAREA_MODE           True
BASEMENTAREA_AVG            True
BASEMENTAREA_MEDI           True
LANDAREA_MEDI               True
LANDAREA_AVG                True
LANDAREA_MODE               True
OWN_CAR_AGE                 True
YEARS_BUILD_MODE            True
YEARS_BUILD_AVG             True
YEARS_BUILD_MEDI            True
FLOORSM

In [ ]:
df_a.shape

(307511, 122)

In [ ]:
df_a.describe().transpose()
#	Target variable 
#1 - client with payment difficulties:he/she had late payment more than X days on at least one of the first Y installments of the loan in our sample,
#0 - all other cases

count           mean            std  \
SK_ID_CURR                    307511.0  278180.518577  102790.175348   
TARGET                        307511.0       0.080729       0.272419   
CNT_CHILDREN                  307511.0       0.417052       0.722121   
AMT_INCOME_TOTAL              307511.0  168797.919297  237123.146279   
AMT_CREDIT                    307511.0  599025.999706  402490.776996   
AMT_ANNUITY                   307499.0   27108.573909   14493.737315   
AMT_GOODS_PRICE               307233.0  538396.207429  369446.460540   
REGION_POPULATION_RELATIVE    307511.0       0.020868       0.013831   
DAYS_BIRTH                    307511.0  -16036.995067    4363.988632   
DAYS_EMPLOYED                 307511.0   63815.045904  141275.766519   
DAYS_REGISTRATION             307511.0   -4986.120328    3522.886321   
DAYS_ID_PUBLISH               307511.0   -2994.202373    1509.450419   
OWN_CAR_AGE                   104582.0      12.061091      11.944812   
FLAG_MOBIL                    307511.0       0.999997       0.001803   
FLAG_EMP_PHONE                307511.0       0.819889       0.384280   
FLAG_WORK_PHONE               307511.0       0.199368       0.399526   
FLAG_CONT_MOBILE              307511.0       0.998133       0.043164   
FLAG_PHONE                    307511.0       0.281066       0.449521   
FLAG_EMAIL                    307511.0       0.056720       0.231307   
CNT_FAM_MEMBERS               307509.0       2.152665       0.910682   
REGION_RATING_CLIENT          307511.0       2.052463       0.509034   
REGION_RATING_CLIENT_W_CITY   307511.0       2.031521       0.502737   
HOUR_APPR_PROCESS_START       307511.0      12.063419       3.265832   
REG_REGION_NOT_LIVE_REGION    307511.0       0.015144       0.122126   
REG_REGION_NOT_WORK_REGION    307511.0       0.050769       0.219526   
LIVE_REGION_NOT_WORK_REGION   307511.0       0.040659       0.197499   
REG_CITY_NOT_LIVE_CITY        307511.0       0.078173       0.268444   
REG_CITY_NOT_WORK_CITY        307511.0       0.230454       0.421124   
LIVE_CITY_NOT_WORK_CITY       307511.0       0.179555       0.383817   
EXT_SOURCE_1                  134133.0       0.502130       0.211062   
EXT_SOURCE_2                  306851.0       0.514393       0.191060   
EXT_SOURCE_3                  246546.0       0.510853       0.194844   
APARTMENTS_AVG                151450.0       0.117440       0.108240   
BASEMENTAREA_AVG              127568.0       0.088442       0.082438   
YEARS_BEGINEXPLUATATION_AVG   157504.0       0.977735       0.059223   
YEARS_BUILD_AVG               103023.0       0.752471       0.113280   
COMMONAREA_AVG                 92646.0       0.044621       0.076036   
ELEVATORS_AVG                 143620.0       0.078942       0.134576   
ENTRANCES_AVG                 152683.0       0.149725       0.100049   
FLOORSMAX_AVG                 154491.0       0.226282       0.144641   
FLOORSMIN_AVG                  98869.0       0.231894       0.161380   
LANDAREA_AVG                  124921.0       0.066333       0.081184   
LIVINGAPARTMENTS_AVG           97312.0       0.100775       0.092576   
LIVINGAREA_AVG                153161.0       0.107399       0.110565   
NONLIVINGAPARTMENTS_AVG        93997.0       0.008809       0.047732   
NONLIVINGAREA_AVG             137829.0       0.028358       0.069523   
APARTMENTS_MODE               151450.0       0.114231       0.107936   
BASEMENTAREA_MODE             127568.0       0.087543       0.084307   
YEARS_BEGINEXPLUATATION_MODE  157504.0       0.977065       0.064575   
YEARS_BUILD_MODE              103023.0       0.759637       0.110111   
COMMONAREA_MODE                92646.0       0.042553       0.074445   
ELEVATORS_MODE                143620.0       0.074490       0.132256   
ENTRANCES_MODE                152683.0       0.145193       0.100977   
FLOORSMAX_MODE                154491.0       0.222315       0.143709   
FLOORSMIN_MODE                 98869.0       0.228058       0.161160   
L

# SUMMARY

1. we have 307511 row and  122 columns

2. 41 columns in dataset are having more than 50% of nan values

3. 8% are the people with payment issue

4. Almost all columns have less standard deviation which means low variance means nearer to mean

5. The data seem to be disbalanced with respect to target variable

if the standard deviation is high that means your data have high variance. And if you compare the mean and 50% percentile (median) you can see the skewness of your numeric variables. From skewness you can infer that your numeric variables have too many outlier or not.


standard deviation----->variance

mean,mode---->skewness

sknewness--->outliers


positively skewed:   mean of the data is greater than the median 

mean=median= mode symmetrical distribution

negatively skewed: mean is less than median


we find skewness of data so that we can transform data in (normal distribution)

In [ ]:
df_a["NAME_CONTRACT_TYPE"].value_counts()

Cash loans         278232
Revolving loans     29279
Name: NAME_CONTRACT_TYPE, dtype: int64

we have 36935 client with cash loan and 3847 with revolving loans

In [ ]:
df_a["TARGET"].value_counts()

0    282686
1     24825
Name: TARGET, dtype: int64

In [ ]:
3299/len(df_a)*100

1.0728071516140885

8% i.e 3299 in the records have payment issue


In [ ]:
df_a["CODE_GENDER"].value_counts()

F      202448
M      105059
XNA         4
Name: CODE_GENDER, dtype: int64

We have 2:1 ratio in our records for F:M 

In [ ]:
df_info=df_a[["AMT_INCOME_TOTAL","AMT_CREDIT"]]
df_info["times"]=np.floor(df_a["AMT_CREDIT"]/df_a["AMT_INCOME_TOTAL"])
df_info.head()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


AMT_INCOME_TOTAL  AMT_CREDIT  times
0          202500.0    406597.5    2.0
1          270000.0   1293502.5    4.0
2           67500.0    135000.0    2.0
3          135000.0    312682.5    2.0
4          121500.0    513000.0    4.0

In [ ]:
#Education based defaluters
df_a[df_a["TARGET"]==1]["NAME_EDUCATION_TYPE"].value_counts()

Secondary / secondary special    19524
Higher education                  4009
Incomplete higher                  872
Lower secondary                    417
Academic degree                      3
Name: NAME_EDUCATION_TYPE, dtype: int64

In [ ]:
#Income type defaluters
df_a[df_a["TARGET"]==1]["NAME_INCOME_TYPE"].value_counts()

Working                 15224
Commercial associate     5360
Pensioner                2982
State servant            1249
Unemployed                  8
Maternity leave             2
Name: NAME_INCOME_TYPE, dtype: int64

# Summary



**To see the distribution of contract type**     

      Contract type       Frequency
      Cash loans           36935
      Revolving loans      3847


**8% i.e 3299 in the records have payment issue**

**female to male ratio is 2:1**

F:      202448

M:      105059

**To see income and credit ratio**

	AMT_INCOME_TOTAL	AMT_CREDIT	times

	202500.0	        406597.5	   2.0

	270000.0	        1293502.5	   4.0

	67500.0	          135000.0	   2.0

	135000.0	        312682.5	   2.0

	121500.0	        513000.0	   4.0



**Education based defaluters**

Secondary / secondary special    19524

Higher education                  4009

Incomplete higher                  872

Lower secondary                    417

Academic degree                      3


**Income type defaluters**

Working                 15224

Commercial associate     5360

Pensioner                2982

State servant            1249

Unemployed                  8

Maternity leave             2


**Insights based on gender**

1: Average income of female is less. But despite that fact ratio of female house owners and male owners is same.

2: Female applicants' average employment days before the application is more than the male applicants.

3: 63% of the loans that were applied by the female applicants was approved(same as male applicants).


4: 72% of the women applicants are from region with the rating 2, same as men.

5: Credit and cash offices are most popular channels through which clients were acquired in the previous application which is 
   common for both the genders.

   



In [ ]:
#Dividing the columns into continuous and discreate column
#threshold =30
cont_col=[]
dis_col=[]
for column in df_a.columns:
  if len(df_a[column].unique())>30:
    cont_col.append(column)
  else:
    dis_col.append(column)

#removing the target variable from dis_col
dis_col.remove("TARGET")

In [ ]:
print(f"No of continuous columns:{len(cont_col)} \n No. of discrete columns {len(dis_col)}")

No of continuous columns:58 
 No. of discrete columns 63


In [ ]:
df_p.head(5)

SK_ID_PREV  SK_ID_CURR NAME_CONTRACT_TYPE  AMT_ANNUITY  AMT_APPLICATION  \
0     2030495      271877     Consumer loans     1730.430          17145.0   
1     2802425      108129         Cash loans    25188.615         607500.0   
2     2523466      122040         Cash loans    15060.735         112500.0   
3     2819243      176158         Cash loans    47041.335         450000.0   
4     1784265      202054         Cash loans    31924.395         337500.0   

   AMT_CREDIT  AMT_DOWN_PAYMENT  AMT_GOODS_PRICE WEEKDAY_APPR_PROCESS_START  \
0     17145.0               0.0          17145.0                   SATURDAY   
1    679671.0               NaN         607500.0                   THURSDAY   
2    136444.5               NaN         112500.0                    TUESDAY   
3    470790.0               NaN         450000.0                     MONDAY   
4    404055.0               NaN         337500.0                   THURSDAY   

   HOUR_APPR_PROCESS_START FLAG_LAST_APPL_PER_CONTRACT  \
0                       15                           Y   
1                       11                           Y   
2                       11                           Y   
3                        7                           Y   
4                        9                           Y   

   NFLAG_LAST_APPL_IN_DAY  RATE_DOWN_PAYMENT  RATE_INTEREST_PRIMARY  \
0                       1                0.0               0.182832   
1                       1                NaN                    NaN   
2                       1                NaN                    NaN   
3                       1                NaN                    NaN   
4                       1                NaN                    NaN   

   RATE_INTEREST_PRIVILEGED NAME_CASH_LOAN_PURPOSE NAME_CONTRACT_STATUS  \
0                  0.867336                    XAP             Approved   
1                       NaN                    XNA             Approved   
2                       NaN                    XNA             Approved   
3                       NaN                    XNA             Approved   
4                       NaN                Repairs              Refused   

   DAYS_DECISION      NAME_PAYMENT_TYPE CODE_REJECT_REASON  NAME_TYPE_SUITE  \
0            -73  Cash through the bank                XAP              NaN   
1           -164                    XNA                XAP    Unaccompanied   
2           -301  Cash through the bank                XAP  Spouse, partner   
3           -512  Cash through the bank                XAP              NaN   
4           -781  Cash through the bank                 HC              NaN   

  NAME_CLIENT_TYPE NAME_GOODS_CATEGORY NAME_PORTFOLIO NAME_PRODUCT_TYPE  \
0         Repeater              Mobile            POS               XNA   
1         Repeater                 XNA           Cash            x-sell   
2         Repeater                 XNA           Cash            x-sell   
3         Repeater                 XNA           Cash            x-sell   
4         Repeater                 XNA           Cash           walk-in   

              CHANNEL_TYPE  SELLERPLACE_AREA NAME_SELLER_INDUSTRY  \
0             Country-wide                35         Connectivity   
1           Contact center                -1                  XNA   
2  Credit and cash offices                -1                  XNA   
3  Credit and cash offices                -1                  XNA   
4  Credit and cash offices                -1                  XNA   

   CNT_PAYMENT NAME_YIELD_GROUP       PRODUCT_COMBINATION  DAYS_FIRST_DRAWING  \
0         12.0           middle  POS mobile with interest            365243.0   
1         36.0       low_action          Cash X-Sell: low            365243.0   
2         12.0             high         Cash X-Sell: high            365243.0   
3         12.0           middle       Cash X-Sell: middle            365243.0   
4         24.0             high         Cash Street: high                 NaN   

   D

In [ ]:
print(f"No of rows and columns are:{df_p.shape}" )

No of rows and columns are:(1670214, 37)


In [ ]:
print("Getting information about column\n\n", df_p.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1670214 entries, 0 to 1670213
Data columns (total 37 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   SK_ID_PREV                   1670214 non-null  int64  
 1   SK_ID_CURR                   1670214 non-null  int64  
 2   NAME_CONTRACT_TYPE           1670214 non-null  object 
 3   AMT_ANNUITY                  1297979 non-null  float64
 4   AMT_APPLICATION              1670214 non-null  float64
 5   AMT_CREDIT                   1670213 non-null  float64
 6   AMT_DOWN_PAYMENT             774370 non-null   float64
 7   AMT_GOODS_PRICE              1284699 non-null  float64
 8   WEEKDAY_APPR_PROCESS_START   1670214 non-null  object 
 9   HOUR_APPR_PROCESS_START      1670214 non-null  int64  
 10  FLAG_LAST_APPL_PER_CONTRACT  1670214 non-null  object 
 11  NFLAG_LAST_APPL_IN_DAY       1670214 non-null  int64  
 12  RATE_DOWN_PAYMENT            774370 non-nu

In [ ]:
print("Getting name of all columns: \n\n",df_p.columns)

Getting name of all columns: 

 Index(['SK_ID_PREV', 'SK_ID_CURR', 'NAME_CONTRACT_TYPE', 'AMT_ANNUITY',
       'AMT_APPLICATION', 'AMT_CREDIT', 'AMT_DOWN_PAYMENT', 'AMT_GOODS_PRICE',
       'WEEKDAY_APPR_PROCESS_START', 'HOUR_APPR_PROCESS_START',
       'FLAG_LAST_APPL_PER_CONTRACT', 'NFLAG_LAST_APPL_IN_DAY',
       'RATE_DOWN_PAYMENT', 'RATE_INTEREST_PRIMARY',
       'RATE_INTEREST_PRIVILEGED', 'NAME_CASH_LOAN_PURPOSE',
       'NAME_CONTRACT_STATUS', 'DAYS_DECISION', 'NAME_PAYMENT_TYPE',
       'CODE_REJECT_REASON', 'NAME_TYPE_SUITE', 'NAME_CLIENT_TYPE',
       'NAME_GOODS_CATEGORY', 'NAME_PORTFOLIO', 'NAME_PRODUCT_TYPE',
       'CHANNEL_TYPE', 'SELLERPLACE_AREA', 'NAME_SELLER_INDUSTRY',
       'CNT_PAYMENT', 'NAME_YIELD_GROUP', 'PRODUCT_COMBINATION',
       'DAYS_FIRST_DRAWING', 'DAYS_FIRST_DUE', 'DAYS_LAST_DUE_1ST_VERSION',
       'DAYS_LAST_DUE', 'DAYS_TERMINATION', 'NFLAG_INSURED_ON_APPROVAL'],
      dtype='object')


In [ ]:
df_p.describe().transpose()

count          mean            std  \
SK_ID_PREV                 1670214.0  1.923089e+06  532597.958696   
SK_ID_CURR                 1670214.0  2.783572e+05  102814.823849   
AMT_ANNUITY                1297979.0  1.595512e+04   14782.137335   
AMT_APPLICATION            1670214.0  1.752339e+05  292779.762387   
AMT_CREDIT                 1670213.0  1.961140e+05  318574.616546   
AMT_DOWN_PAYMENT            774370.0  6.697402e+03   20921.495410   
AMT_GOODS_PRICE            1284699.0  2.278473e+05  315396.557937   
HOUR_APPR_PROCESS_START    1670214.0  1.248418e+01       3.334028   
NFLAG_LAST_APPL_IN_DAY     1670214.0  9.964675e-01       0.059330   
RATE_DOWN_PAYMENT           774370.0  7.963682e-02       0.107823   
RATE_INTEREST_PRIMARY         5951.0  1.883569e-01       0.087671   
RATE_INTEREST_PRIVILEGED      5951.0  7.735025e-01       0.100879   
DAYS_DECISION              1670214.0 -8.806797e+02     779.099667   
SELLERPLACE_AREA           1670214.0  3.139511e+02    7127.443459   
CNT_PAYMENT                1297984.0  1.605408e+01      14.567288   
DAYS_FIRST_DRAWING          997149.0  3.422099e+05   88916.115834   
DAYS_FIRST_DUE              997149.0  1.382627e+04   72444.869708   
DAYS_LAST_DUE_1ST_VERSION   997149.0  3.376777e+04  106857.034789   
DAYS_LAST_DUE               997149.0  7.658240e+04  149647.415123   
DAYS_TERMINATION            997149.0  8.199234e+04  153303.516729   
NFLAG_INSURED_ON_APPROVAL   997149.0  3.325702e-01       0.471134   

                                    min           25%           50%  \
SK_ID_PREV                 1.000001e+06  1.461857e+06  1.923110e+06   
SK_ID_CURR                 1.000010e+05  1.893290e+05  2.787145e+05   
AMT_ANNUITY                0.000000e+00  6.321780e+03  1.125000e+04   
AMT_APPLICATION            0.000000e+00  1.872000e+04  7.104600e+04   
AMT_CREDIT                 0.000000e+00  2.416050e+04  8.054100e+04   
AMT_DOWN_PAYMENT          -9.000000e-01  0.000000e+00  1.638000e+03   
AMT_GOODS_PRICE            0.000000e+00  5.084100e+04  1.123200e+05   
HOUR_APPR_PROCESS_START    0.000000e+00  1.000000e+01  1.200000e+01   
NFLAG_LAST_APPL_IN_DAY     0.000000e+00  1.000000e+00  1.000000e+00   
RATE_DOWN_PAYMENT         -1.497876e-05  0.000000e+00  5.160508e-02   
RATE_INTEREST_PRIMARY      3.478125e-02  1.607163e-01  1.891222e-01   
RATE_INTEREST_PRIVILEGED   3.731501e-01  7.156448e-01  8.350951e-01   
DAYS_DECISION             -2.922000e+03 -1.300000e+03 -5.810000e+02   
SELLERPLACE_AREA          -1.000000e+00 -1.000000e+00  3.000000e+00   
CNT_PAYMENT                0.000000e+00  6.000000e+00  1.200000e+01   
DAYS_FIRST_DRAWING        -2.922000e+03  3.652430e+05  3.652430e+05   
DAYS_FIRST_DUE            -2.892000e+03 -1.628000e+03 -8.310000e+02   
DAYS_LAST_DUE_1ST_VERSION -2.801000e+03 -1.242000e+03 -3.610000e+02   
DAYS_LAST_DUE             -2.889000e+03 -1.314000e+03 -5.370000e+02   
DAYS_TERMINATION          -2.874000e+03 -1.270000e+03 -4.990000e+02   
NFLAG_INSURED_ON_APPROVAL  0.000000e+00  0.000000e+00  0.000000e+00   

                                    75%          max  
SK_ID_PREV                 2.384280e+06  2845382.000  
SK_ID_CURR                 3.675140e+05   456255.000  
AMT_ANNUITY                2.065842e+04   418058.145  
AMT_APPLICATION            1.803600e+05  6905160.000  
AMT_CREDIT                 2.164185e+05  6905160.000  
AMT_DOWN_PAYMENT           7.740000e+03  3060045.000  
AMT_GOODS_PRICE            2.340000e+05  6905160.000  
HOUR_APPR_PROCESS_START    1.500000e+01       23.000  
NFLAG_LAST_APPL_IN_DAY     1.000000e+00        1.000  
RATE_DOWN_PAYMENT          1.089091e-01        1.000  
RATE_INTEREST_PRIMARY      1.933299e-01        1.000  
RATE_INTEREST_PRIVILEGED   8.525370e-01        1.000  
DAYS_DECISION             -2.800000e+02       -1.000  
SELLERPLACE_AREA           8.200000e+01  4000000.000  
CNT_PAYMENT                2.400000e+01       84.000  
DAYS_FIRST_DRAWING         3.652430e+05   365243.000  
DAYS_FIRST_DUE

In [ ]:
(df_p.isna().sum()/len(df_p)*100).sort_values()

SK_ID_PREV                      0.000000
NAME_YIELD_GROUP                0.000000
NAME_SELLER_INDUSTRY            0.000000
SELLERPLACE_AREA                0.000000
CHANNEL_TYPE                    0.000000
NAME_PRODUCT_TYPE               0.000000
NAME_PORTFOLIO                  0.000000
NAME_GOODS_CATEGORY             0.000000
NAME_CLIENT_TYPE                0.000000
CODE_REJECT_REASON              0.000000
DAYS_DECISION                   0.000000
NAME_CONTRACT_STATUS            0.000000
NAME_CASH_LOAN_PURPOSE          0.000000
NAME_PAYMENT_TYPE               0.000000
AMT_APPLICATION                 0.000000
NFLAG_LAST_APPL_IN_DAY          0.000000
FLAG_LAST_APPL_PER_CONTRACT     0.000000
HOUR_APPR_PROCESS_START         0.000000
WEEKDAY_APPR_PROCESS_START      0.000000
NAME_CONTRACT_TYPE              0.000000
SK_ID_CURR                      0.000000
AMT_CREDIT                      0.000060
PRODUCT_COMBINATION             0.020716
CNT_PAYMENT                    22.286366
AMT_ANNUITY     

In [ ]:
#correlation between df(application) and df_p(previous application)
col_app=list(df_a.columns)
col_prev=list(df_p.columns)
uni_col=[]
[uni_col.append(col) for col  in col_app if col not in uni_col]
[uni_col.append(col) for col in col_prev if col not in col_app]

uni_col=list(set(uni_col))
    
print(f"List of columns unique in application.csv and previous application:{uni_col}")

List of columns unique in application.csv and previous application:['NONLIVINGAREA_AVG', 'ENTRANCES_AVG', 'NAME_HOUSING_TYPE', 'AMT_REQ_CREDIT_BUREAU_QRT', 'REG_CITY_NOT_LIVE_CITY', 'SELLERPLACE_AREA', 'FLAG_LAST_APPL_PER_CONTRACT', 'ORGANIZATION_TYPE', 'FONDKAPREMONT_MODE', 'REG_REGION_NOT_LIVE_REGION', 'AMT_REQ_CREDIT_BUREAU_DAY', 'FLAG_DOCUMENT_15', 'AMT_REQ_CREDIT_BUREAU_MON', 'HOUR_APPR_PROCESS_START', 'WALLSMATERIAL_MODE', 'FLAG_DOCUMENT_10', 'AMT_INCOME_TOTAL', 'FLAG_PHONE', 'TARGET', 'CNT_CHILDREN', 'APARTMENTS_MEDI', 'TOTALAREA_MODE', 'RATE_INTEREST_PRIVILEGED', 'CHANNEL_TYPE', 'NAME_INCOME_TYPE', 'FLAG_DOCUMENT_18', 'AMT_REQ_CREDIT_BUREAU_HOUR', 'SK_ID_CURR', 'APARTMENTS_MODE', 'NAME_PRODUCT_TYPE', 'AMT_APPLICATION', 'DAYS_LAST_DUE', 'NAME_PORTFOLIO', 'WEEKDAY_APPR_PROCESS_START', 'EXT_SOURCE_1', 'LIVINGAPARTMENTS_MEDI', 'CODE_GENDER', 'FLAG_DOCUMENT_8', 'NFLAG_LAST_APPL_IN_DAY', 'NONLIVINGAPARTMENTS_MEDI', 'YEARS_BUILD_MEDI', 'NAME_TYPE_SUITE', 'CNT_PAYMENT', 'REGION_RATING_

In [ ]:
#common columns
common=[]
[common.append(col) for col in col_app if col in col_prev]
common

['SK_ID_CURR',
 'NAME_CONTRACT_TYPE',
 'AMT_CREDIT',
 'AMT_ANNUITY',
 'AMT_GOODS_PRICE',
 'NAME_TYPE_SUITE',
 'WEEKDAY_APPR_PROCESS_START',
 'HOUR_APPR_PROCESS_START']

In [ ]:
#merging the two dataset
final_df=pd.merge(df_a, df_p,on='SK_ID_CURR')

In [ ]:
#final_df.head()

In [ ]:
final_df.shape

NameError: ignored

In [ ]:
lst=list(final_df.columns)
len(lst)

158

In [ ]:
lst.index("SK_ID_PREV")

122

In [ ]:
col_lst=[]
col_lst.append(lst[122])
col_lst.extend(lst[:122])
col_lst.extend(lst[123:])
col_lst

['SK_ID_PREV',
 'SK_ID_CURR',
 'TARGET',
 'NAME_CONTRACT_TYPE_x',
 'CODE_GENDER',
 'FLAG_OWN_CAR',
 'FLAG_OWN_REALTY',
 'CNT_CHILDREN',
 'AMT_INCOME_TOTAL',
 'AMT_CREDIT_x',
 'AMT_ANNUITY_x',
 'AMT_GOODS_PRICE_x',
 'NAME_TYPE_SUITE_x',
 'NAME_INCOME_TYPE',
 'NAME_EDUCATION_TYPE',
 'NAME_FAMILY_STATUS',
 'NAME_HOUSING_TYPE',
 'REGION_POPULATION_RELATIVE',
 'DAYS_BIRTH',
 'DAYS_EMPLOYED',
 'DAYS_REGISTRATION',
 'DAYS_ID_PUBLISH',
 'OWN_CAR_AGE',
 'FLAG_MOBIL',
 'FLAG_EMP_PHONE',
 'FLAG_WORK_PHONE',
 'FLAG_CONT_MOBILE',
 'FLAG_PHONE',
 'FLAG_EMAIL',
 'OCCUPATION_TYPE',
 'CNT_FAM_MEMBERS',
 'REGION_RATING_CLIENT',
 'REGION_RATING_CLIENT_W_CITY',
 'WEEKDAY_APPR_PROCESS_START_x',
 'HOUR_APPR_PROCESS_START_x',
 'REG_REGION_NOT_LIVE_REGION',
 'REG_REGION_NOT_WORK_REGION',
 'LIVE_REGION_NOT_WORK_REGION',
 'REG_CITY_NOT_LIVE_CITY',
 'REG_CITY_NOT_WORK_CITY',
 'LIVE_CITY_NOT_WORK_CITY',
 'ORGANIZATION_TYPE',
 'EXT_SOURCE_1',
 'EXT_SOURCE_2',
 'EXT_SOURCE_3',
 'APARTMENTS_AVG',
 'BASEMENTAREA_AVG'

In [ ]:
len(col_lst)

158

In [ ]:
df=final_df[col_lst]
df.head()

SK_ID_PREV  SK_ID_CURR  TARGET NAME_CONTRACT_TYPE_x CODE_GENDER  \
0     1038818      100002       1           Cash loans           M   
1     1810518      100003       0           Cash loans           F   
2     2636178      100003       0           Cash loans           F   
3     2396755      100003       0           Cash loans           F   
4     1564014      100004       0      Revolving loans           M   

  FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT_x  \
0            N               Y             0          202500.0      406597.5   
1            N               N             0          270000.0     1293502.5   
2            N               N             0          270000.0     1293502.5   
3            N               N             0          270000.0     1293502.5   
4            Y               Y             0           67500.0      135000.0   

   AMT_ANNUITY_x  AMT_GOODS_PRICE_x NAME_TYPE_SUITE_x NAME_INCOME_TYPE  \
0        24700.5           351000.0     Unaccompanied          Working   
1        35698.5          1129500.0            Family    State servant   
2        35698.5          1129500.0            Family    State servant   
3        35698.5          1129500.0            Family    State servant   
4         6750.0           135000.0     Unaccompanied          Working   

             NAME_EDUCATION_TYPE    NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  \
0  Secondary / secondary special  Single / not married  House / apartment   
1               Higher education               Married  House / apartment   
2               Higher education               Married  House / apartment   
3               Higher education               Married  House / apartment   
4  Secondary / secondary special  Single / not married  House / apartment   

   REGION_POPULATION_RELATIVE  DAYS_BIRTH  DAYS_EMPLOYED  DAYS_REGISTRATION  \
0                    0.018801       -9461           -637            -3648.0   
1                    0.003541      -16765          -1188            -1186.0   
2                    0.003541      -16765          -1188            -1186.0   
3                    0.003541      -16765          -1188            -1186.0   
4                    0.010032      -19046           -225            -4260.0   

   DAYS_ID_PUBLISH  OWN_CAR_AGE  FLAG_MOBIL  FLAG_EMP_PHONE  FLAG_WORK_PHONE  \
0            -2120          NaN           1               1                0   
1             -291          NaN           1               1                0   
2             -291          NaN           1               1                0   
3             -291          NaN           1               1                0   
4            -2531         26.0           1               1                1   

   FLAG_CONT_MOBILE  FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  CNT_FAM_MEMBERS  \
0                 1           1           0        Laborers              1.0   
1                 1           1           0      Core staff              2.0   
2                 1           1           0      Core staff              2.0   
3                 1           1           0      Core staff              2.0   
4                 1           1           0        Laborers              1.0   

   REGION_RATING_CLIENT  REGION_RATING_CLIENT_W_CITY  \
0                     2                            2   
1                     1                            1   
2                     1                            1   
3                     1                            1   
4                     2                            2   

  WEEKDAY_APPR_PROCESS_START_x  HOUR_APPR_PROCESS_START_x  \
0                    WEDNESDAY                         10   
1                       MONDAY                         11   
2                       MONDAY                         11   
3                       MONDAY                         11   
4                       MONDAY                          9   

   REG_REGION_NOT_LIVE_REGION  REG_REGION_NOT_WORK_R

In [ ]:
df.shape

(1413701, 158)

We have 1413701 rows and 158 columns after merging the data from application.csv and prev_application.

In [ ]:
df["TARGET"].value_counts()

0    1291341
1     122360
Name: TARGET, dtype: int64

The dataset seems to be disbalanced with respective to target variable.

In [ ]:
#columns with more than 50% nan values
print("columns with more than 50 % of nan values")
nans=((df.isna().sum()/len(df))*100).sort_values()
print("count: ",nans[nans.values>50].count())
nans[nans.values>50]

columns with more than 50 % of nan values
count:  38


APARTMENTS_MODE             50.364115
APARTMENTS_AVG              50.364115
APARTMENTS_MEDI             50.364115
WALLSMATERIAL_MODE          50.462297
ELEVATORS_AVG               52.996779
ELEVATORS_MODE              52.996779
ELEVATORS_MEDI              52.996779
AMT_DOWN_PAYMENT            53.019698
RATE_DOWN_PAYMENT           53.019698
NONLIVINGAREA_AVG           54.860115
NONLIVINGAREA_MEDI          54.860115
NONLIVINGAREA_MODE          54.860115
EXT_SOURCE_1                56.366870
BASEMENTAREA_AVG            58.308511
BASEMENTAREA_MEDI           58.308511
BASEMENTAREA_MODE           58.308511
LANDAREA_MEDI               59.218533
LANDAREA_MODE               59.218533
LANDAREA_AVG                59.218533
OWN_CAR_AGE                 66.294499
YEARS_BUILD_MODE            66.298107
YEARS_BUILD_AVG             66.298107
YEARS_BUILD_MEDI            66.298107
FLOORSMIN_AVG               67.670604
FLOORSMIN_MODE              67.670604
FLOORSMIN_MEDI              67.670604
LIVINGAPARTM


38 columns with  more than 50%  nan values are dependent on each other in which Rate_interest_Privileged and Rate_interest_Primay with 99% nan value 

In [ ]:
df.describe().transpose()

count          mean            std  \
SK_ID_PREV                    1413701.0  1.922744e+06  532715.335255   
SK_ID_CURR                    1413701.0  2.784813e+05  102811.788043   
TARGET                        1413701.0  8.655296e-02       0.281179   
CNT_CHILDREN                  1413701.0  4.048933e-01       0.717345   
AMT_INCOME_TOTAL              1413701.0  1.733160e+05  198573.372016   
AMT_CREDIT_x                  1413701.0  5.875537e+05  384917.331385   
AMT_ANNUITY_x                 1413608.0  2.701702e+04   13951.164803   
AMT_GOODS_PRICE_x             1412493.0  5.277186e+05  353246.491860   
REGION_POPULATION_RELATIVE    1413701.0  2.074985e-02       0.013347   
DAYS_BIRTH                    1413701.0 -1.632105e+04    4344.557089   
DAYS_EMPLOYED                 1413701.0  6.870203e+04  145277.704778   
DAYS_REGISTRATION             1413701.0 -5.003233e+03    3551.050912   
DAYS_ID_PUBLISH               1413701.0 -3.034563e+03    1507.376375   
OWN_CAR_AGE                    476495.0  1.201195e+01      12.067141   
FLAG_MOBIL                    1413701.0  1.000000e+00       0.000000   
FLAG_EMP_PHONE                1413701.0  8.064393e-01       0.395089   
FLAG_WORK_PHONE               1413701.0  1.822585e-01       0.386058   
FLAG_CONT_MOBILE              1413701.0  9.992452e-01       0.027462   
FLAG_PHONE                    1413701.0  2.828752e-01       0.450396   
FLAG_EMAIL                    1413701.0  7.242904e-02       0.259197   
CNT_FAM_MEMBERS               1413701.0  2.150501e+00       0.900679   
REGION_RATING_CLIENT          1413701.0  2.059638e+00       0.495780   
REGION_RATING_CLIENT_W_CITY   1413701.0  2.037603e+00       0.488643   
HOUR_APPR_PROCESS_START_x     1413701.0  1.198433e+01       3.232181   
REG_REGION_NOT_LIVE_REGION    1413701.0  1.207327e-02       0.109213   
REG_REGION_NOT_WORK_REGION    1413701.0  4.396686e-02       0.205021   
LIVE_REGION_NOT_WORK_REGION   1413701.0  3.604793e-02       0.186410   
REG_CITY_NOT_LIVE_CITY        1413701.0  7.360538e-02       0.261128   
REG_CITY_NOT_WORK_CITY        1413701.0  2.221481e-01       0.415690   
LIVE_CITY_NOT_WORK_CITY       1413701.0  1.740601e-01       0.379161   
EXT_SOURCE_1                   616842.0  5.054990e-01       0.210748   
EXT_SOURCE_2                  1411880.0  5.114275e-01       0.191119   
EXT_SOURCE_3                  1171859.0  4.896986e-01       0.196536   
APARTMENTS_AVG                 701703.0  1.149783e-01       0.104870   
BASEMENTAREA_AVG               589393.0  8.698938e-02       0.080745   
YEARS_BEGINEXPLUATATION_AVG    729019.0  9.772352e-01       0.061647   
YEARS_BUILD_AVG                476444.0  7.496433e-01       0.113048   
COMMONAREA_AVG                 428502.0  4.340559e-02       0.073492   
ELEVATORS_AVG                  664485.0  7.525023e-02       0.130434   
ENTRANCES_AVG                  706919.0  1.487248e-01       0.099005   
FLOORSMAX_AVG                  715193.0  2.223169e-01       0.140943   
FLOORSMIN_AVG                  457041.0  2.287916e-01       0.158112   
LANDAREA_AVG                   576528.0  6.561416e-02       0.079438   
LIVINGAPARTMENTS_AVG           449351.0  9.886160e-02       0.090477   
LIVINGAREA_AVG                 709301.0  1.047002e-01       0.107025   
NONLIVINGAPARTMENTS_AVG        434055.0  8.608387e-03       0.047742   
NONLIVINGAREA_AVG              638143.0  2.747308e-02       0.067045   
APARTMENTS_MODE                701703.0  1.120754e-01       0.104846   
BASEMENTAREA_MODE              589393.0  8.627123e-02       0.082634   
YEARS_BEGINEXPLUATATION_MODE   729019.0  9.765155e-01       0.067386   
YEARS_BUILD_MODE               476444.0  7.570512e-01       0.109991   
COMMONAREA_MODE                428502.0  4.153704e-02       0.072112   
ELEVATORS_MODE                 664485.0  7.115661e-02       0.128482   
ENTRANCES_MODE                 706919.0  1.444221e-01       0.099944   
FLOORSMAX_MODE                 715193.0  2.185555e-01       0.140270   
FL

In [ ]:
#columns which are similar with 2 dataset after merging
import re 
for column in df.columns:
  if re.search(".*_x$",str(column))!=None:
    print(re.search(".*_x$",str(column)).group())



NAME_CONTRACT_TYPE_x
AMT_CREDIT_x
AMT_ANNUITY_x
AMT_GOODS_PRICE_x
NAME_TYPE_SUITE_x
WEEKDAY_APPR_PROCESS_START_x
HOUR_APPR_PROCESS_START_x


# **SUMMARY**

1. We have 1413701 rows and 158 columns after merging the data from application.csv and prev_application.


2. 38 columns with  more than 50%  nan values are dependent on each other in which Rate_interest_Privileged and Rate_interest_Primay with 99% nan value 

3. We have 8% of defaulter i.e with payment issue according to the target variable.

4. In most of the cases we have low standard deviation which will tell us less variance  and values closer to mean which is good for classification

5. The dataset seems to be disbalanced with respective to target variable (92% 8% split.)

6. positively skewed data in most of our columns

7. Columns which are similar after merging
  
  NAME_CONTRACT_TYPE_x
  
  AMT_CREDIT_x

  AMT_ANNUITY_x

  AMT_GOODS_PRICE_x

  NAME_TYPE_SUITE_x

  WEEKDAY_APPR_PROCESS_START_x

  HOUR_APPR_PROCESS_START_x


8. Duplicate values are present after merging of data.

9. Source of income

    Working                 724089

    Commercial associate    321481

    Pensioner               273544

    State servant            94424

    Unemployed                 123

    Student                     24

    Maternity leave             16

**Create hypothesis if required.**

**Hypothesis are made at the preprocessing and eda state and further they are made more accurate to business problem at the ending stage.**

In [ ]:
df.head()

SK_ID_PREV  SK_ID_CURR  TARGET NAME_CONTRACT_TYPE_x CODE_GENDER  \
0     1038818      100002       1           Cash loans           M   
1     1810518      100003       0           Cash loans           F   
2     2636178      100003       0           Cash loans           F   
3     2396755      100003       0           Cash loans           F   
4     1564014      100004       0      Revolving loans           M   

  FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT_x  \
0            N               Y             0          202500.0      406597.5   
1            N               N             0          270000.0     1293502.5   
2            N               N             0          270000.0     1293502.5   
3            N               N             0          270000.0     1293502.5   
4            Y               Y             0           67500.0      135000.0   

   AMT_ANNUITY_x  AMT_GOODS_PRICE_x NAME_TYPE_SUITE_x NAME_INCOME_TYPE  \
0        24700.5           351000.0     Unaccompanied          Working   
1        35698.5          1129500.0            Family    State servant   
2        35698.5          1129500.0            Family    State servant   
3        35698.5          1129500.0            Family    State servant   
4         6750.0           135000.0     Unaccompanied          Working   

             NAME_EDUCATION_TYPE    NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  \
0  Secondary / secondary special  Single / not married  House / apartment   
1               Higher education               Married  House / apartment   
2               Higher education               Married  House / apartment   
3               Higher education               Married  House / apartment   
4  Secondary / secondary special  Single / not married  House / apartment   

   REGION_POPULATION_RELATIVE  DAYS_BIRTH  DAYS_EMPLOYED  DAYS_REGISTRATION  \
0                    0.018801       -9461           -637            -3648.0   
1                    0.003541      -16765          -1188            -1186.0   
2                    0.003541      -16765          -1188            -1186.0   
3                    0.003541      -16765          -1188            -1186.0   
4                    0.010032      -19046           -225            -4260.0   

   DAYS_ID_PUBLISH  OWN_CAR_AGE  FLAG_MOBIL  FLAG_EMP_PHONE  FLAG_WORK_PHONE  \
0            -2120          NaN           1               1                0   
1             -291          NaN           1               1                0   
2             -291          NaN           1               1                0   
3             -291          NaN           1               1                0   
4            -2531         26.0           1               1                1   

   FLAG_CONT_MOBILE  FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  CNT_FAM_MEMBERS  \
0                 1           1           0        Laborers              1.0   
1                 1           1           0      Core staff              2.0   
2                 1           1           0      Core staff              2.0   
3                 1           1           0      Core staff              2.0   
4                 1           1           0        Laborers              1.0   

   REGION_RATING_CLIENT  REGION_RATING_CLIENT_W_CITY  \
0                     2                            2   
1                     1                            1   
2                     1                            1   
3                     1                            1   
4                     2                            2   

  WEEKDAY_APPR_PROCESS_START_x  HOUR_APPR_PROCESS_START_x  \
0                    WEDNESDAY                         10   
1                       MONDAY                         11   
2                       MONDAY                         11   
3                       MONDAY                         11   
4                       MONDAY                          9   

   REG_REGION_NOT_LIVE_REGION  REG_REGION_NOT_WORK_R

In [ ]:
sim_df=df[["SK_ID_PREV","SK_ID_CURR","NAME_CONTRACT_TYPE_x","NAME_CONTRACT_TYPE_y",
        "AMT_CREDIT_x","AMT_CREDIT_y","AMT_ANNUITY_x","AMT_ANNUITY_y","AMT_GOODS_PRICE_x","AMT_GOODS_PRICE_y",
        "NAME_TYPE_SUITE_x","NAME_TYPE_SUITE_y","WEEKDAY_APPR_PROCESS_START_y","HOUR_APPR_PROCESS_START_y"]]
sim_df.head(10)

SK_ID_PREV  SK_ID_CURR NAME_CONTRACT_TYPE_x NAME_CONTRACT_TYPE_y  \
0     1038818      100002           Cash loans       Consumer loans   
1     1810518      100003           Cash loans           Cash loans   
2     2636178      100003           Cash loans       Consumer loans   
3     2396755      100003           Cash loans       Consumer loans   
4     1564014      100004      Revolving loans       Consumer loans   
5     2078043      100006           Cash loans           Cash loans   
6     2827850      100006           Cash loans      Revolving loans   
7     2190416      100006           Cash loans       Consumer loans   
8     1489396      100006           Cash loans      Revolving loans   
9     1020698      100006           Cash loans           Cash loans   

   AMT_CREDIT_x  AMT_CREDIT_y  AMT_ANNUITY_x  AMT_ANNUITY_y  \
0      406597.5      179055.0        24700.5       9251.775   
1     1293502.5     1035882.0        35698.5      98356.995   
2     1293502.5      348637.5        35698.5      64567.665   
3     1293502.5       68053.5        35698.5       6737.310   
4      135000.0       20106.0         6750.0       5357.250   
5      312682.5      675000.0        29686.5      24246.000   
6      312682.5           0.0        29686.5            NaN   
7      312682.5      267930.0        29686.5      29027.520   
8      312682.5      270000.0        29686.5      13500.000   
9      312682.5      481495.5        29686.5      39954.510   

   AMT_GOODS_PRICE_x  AMT_GOODS_PRICE_y NAME_TYPE_SUITE_x NAME_TYPE_SUITE_y  \
0           351000.0           179055.0     Unaccompanied               NaN   
1          1129500.0           900000.0            Family     Unaccompanied   
2          1129500.0           337500.0            Family            Family   
3          1129500.0            68809.5            Family            Family   
4           135000.0            24282.0     Unaccompanied     Unaccompanied   
5           297000.0           675000.0     Unaccompanied     Unaccompanied   
6           297000.0                NaN     Unaccompanied               NaN   
7           297000.0           334917.0     Unaccompanied            Family   
8           297000.0           270000.0     Unaccompanied     Unaccompanied   
9           297000.0           454500.0     Unaccompanied               NaN   

  WEEKDAY_APPR_PROCESS_START_y  HOUR_APPR_PROCESS_START_y  
0                     SATURDAY                          9  
1                       FRIDAY                         12  
2                       SUNDAY                         17  
3                     SATURDAY                         15  
4                       FRIDAY                          5  
5                     THURSDAY                         15  
6                     THURSDAY                         15  
7                       SUNDAY                         15  
8                     THURSDAY                         15  
9                     SATURDAY                         12



1. Removing duplicate rows values by converting it in single row 
2. Removing duplicate columns values by converting it in single column




Idea for single row

    cur_id    pre_id

    100006     [1564014,2015477,565015]

Idea for single column

    [x,y]

    x---> present

    y--->previous

In [ ]:
#creating a group based on column SK_ID_CURR 
sim_group=sim_df.groupby(by="SK_ID_CURR",sort=True)
#creating simlar column in consideration
similar_list=["SK_ID_PREV","NAME_CONTRACT_TYPE_y","AMT_CREDIT_y","AMT_ANNUITY_y","AMT_GOODS_PRICE_y","NAME_TYPE_SUITE_y","WEEKDAY_APPR_PROCESS_START_y","HOUR_APPR_PROCESS_START_y"]
sim_df_list=pd.DataFrame()

for column in similar_list:
  sim_df_list[column[:len(column)-1]+"list"]=sim_group[column].apply(list)

display(sim_df_list.head(5))


SK_ID_PRElist  \
SK_ID_CURR                                                      
100002                                              [1038818]   
100003                            [1810518, 2636178, 2396755]   
100004                                              [1564014]   
100006      [2078043, 2827850, 2190416, 1489396, 1020698, ...   
100007      [2001242, 2730157, 1692033, 2119973, 2462742, ...   

                                      NAME_CONTRACT_TYPE_list  \
SK_ID_CURR                                                      
100002                                       [Consumer loans]   
100003           [Cash loans, Consumer loans, Consumer loans]   
100004                                       [Consumer loans]   
100006      [Cash loans, Revolving loans, Consumer loans, ...   
100007      [Cash loans, Cash loans, Cash loans, Consumer ...   

                                              AMT_CREDIT_list  \
SK_ID_CURR                                                      
100002                                             [179055.0]   
100003                         [1035882.0, 348637.5, 68053.5]   
100004                                              [20106.0]   
100006      [675000.0, 0.0, 267930.0, 270000.0, 481495.5, ...   
100007      [180000.0, 284400.0, 274288.5, 14616.0, 30663....   

                                             AMT_ANNUITY_list  \
SK_ID_CURR                                                      
100002                                             [9251.775]   
100003                        [98356.995, 64567.665, 6737.31]   
100004                                              [5357.25]   
100006      [24246.0, nan, 29027.52, 13500.0, 39954.51, na...   
100007      [16509.6, 13010.985, 16037.64, 1834.29, 3601.5...   

                                         AMT_GOODS_PRICE_list  \
SK_ID_CURR                                                      
100002                                             [179055.0]   
100003                          [900000.0, 337500.0, 68809.5]   
100004                                              [24282.0]   
100006      [675000.0, nan, 334917.0, 270000.0, 454500.0, ...   
100007      [180000.0, 225000.0, 247500.0, 17176.5, 31005....   

                                         NAME_TYPE_SUITE_list  \
SK_ID_CURR                                                      
100002                                                  [nan]   
100003                        [Unaccompanied, Family, Family]   
100004                                        [Unaccompanied]   
100006      [Unaccompanied, nan, Family, Unaccompanied, na...   
100007      [Unaccompanied, Family, nan, Family, Unaccompa...   

                              WEEKDAY_APPR_PROCESS_START_list  \
SK_ID_CURR                                                      
100002                                             [SATURDAY]   
100003                             [FRIDAY, SUNDAY, SATURDAY]   
100004                                               [FRIDAY]   
100006      [THURSDAY, THURSDAY, SUNDAY, THURSDAY, SATURDA...   
100007      [SUNDAY, FRIDAY, MONDAY, SATURDAY, SUNDAY, THU...   

                    HOUR_APPR_PROCESS_START_list  
SK_ID_CURR                                        
100002                                       [9]  
100003                              [12, 17, 15]  
100004                                       [5]  
100006      [15, 15, 15, 15, 12, 15, 15, 15, 15]  
100007                   [14, 13, 11, 13, 15, 8]

**Group by helps us in two ways one is transformation and second is applying aggregate function a group of groups.**

In [ ]:
lst=["SK_ID_PREV","NAME_CONTRACT_TYPE","AMT_CREDIT","AMT_ANNUITY","AMT_GOODS_PRICE","NAME_TYPE_SUITE","WEEKDAY_APPR_PROCESS_START","HOUR_APPR_PROCESS_START"]
df_p.drop(lst,axis=1,inplace=True)


In [ ]:
display(df_p.head())
#df_p.shape

SK_ID_CURR  AMT_APPLICATION  AMT_DOWN_PAYMENT FLAG_LAST_APPL_PER_CONTRACT  \
0      271877          17145.0               0.0                           Y   
1      108129         607500.0               NaN                           Y   
2      122040         112500.0               NaN                           Y   
3      176158         450000.0               NaN                           Y   
4      202054         337500.0               NaN                           Y   

   NFLAG_LAST_APPL_IN_DAY  RATE_DOWN_PAYMENT  RATE_INTEREST_PRIMARY  \
0                       1                0.0               0.182832   
1                       1                NaN                    NaN   
2                       1                NaN                    NaN   
3                       1                NaN                    NaN   
4                       1                NaN                    NaN   

   RATE_INTEREST_PRIVILEGED NAME_CASH_LOAN_PURPOSE NAME_CONTRACT_STATUS  \
0                  0.867336                    XAP             Approved   
1                       NaN                    XNA             Approved   
2                       NaN                    XNA             Approved   
3                       NaN                    XNA             Approved   
4                       NaN                Repairs              Refused   

   DAYS_DECISION      NAME_PAYMENT_TYPE CODE_REJECT_REASON NAME_CLIENT_TYPE  \
0            -73  Cash through the bank                XAP         Repeater   
1           -164                    XNA                XAP         Repeater   
2           -301  Cash through the bank                XAP         Repeater   
3           -512  Cash through the bank                XAP         Repeater   
4           -781  Cash through the bank                 HC         Repeater   

  NAME_GOODS_CATEGORY NAME_PORTFOLIO NAME_PRODUCT_TYPE  \
0              Mobile            POS               XNA   
1                 XNA           Cash            x-sell   
2                 XNA           Cash            x-sell   
3                 XNA           Cash            x-sell   
4                 XNA           Cash           walk-in   

              CHANNEL_TYPE  SELLERPLACE_AREA NAME_SELLER_INDUSTRY  \
0             Country-wide                35         Connectivity   
1           Contact center                -1                  XNA   
2  Credit and cash offices                -1                  XNA   
3  Credit and cash offices                -1                  XNA   
4  Credit and cash offices                -1                  XNA   

   CNT_PAYMENT NAME_YIELD_GROUP       PRODUCT_COMBINATION  DAYS_FIRST_DRAWING  \
0         12.0           middle  POS mobile with interest            365243.0   
1         36.0       low_action          Cash X-Sell: low            365243.0   
2         12.0             high         Cash X-Sell: high            365243.0   
3         12.0           middle       Cash X-Sell: middle            365243.0   
4         24.0             high         Cash Street: high                 NaN   

   DAYS_FIRST_DUE  DAYS_LAST_DUE_1ST_VERSION  DAYS_LAST_DUE  DAYS_TERMINATION  \
0           -42.0                      300.0          -42.0             -37.0   
1          -134.0                      916.0       365243.0          365243.0   
2          -271.0                       59.0       365243.0          365243.0   
3          -482.0                     -152.0         -182.0            -177.0   
4             NaN                        NaN            NaN               NaN   

   NFLAG_INSURED_ON_APPROVAL  
0                        0.0  
1                        1.0  
2                        1.0  
3                        1.0  
4                        NaN

In [ ]:
#merging the list data
df_m=pd.merge(df_a,df_p,on="SK_ID_CURR")

#df.head()

In [ ]:
#df.shape

In [ ]:
df=pd.merge(df_m,sim_df_list,on="SK_ID_CURR")

In [ ]:
#df.head()

#

In [ ]:
df.shape

(1413701, 158)

In [ ]:
df.head()

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
0      100002       1         Cash loans           M            N   
1      100003       0         Cash loans           F            N   
2      100003       0         Cash loans           F            N   
3      100003       0         Cash loans           F            N   
4      100004       0    Revolving loans           M            Y   

  FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0               Y             0          202500.0    406597.5      24700.5   
1               N             0          270000.0   1293502.5      35698.5   
2               N             0          270000.0   1293502.5      35698.5   
3               N             0          270000.0   1293502.5      35698.5   
4               Y             0           67500.0    135000.0       6750.0   

   AMT_GOODS_PRICE NAME_TYPE_SUITE NAME_INCOME_TYPE  \
0         351000.0   Unaccompanied          Working   
1        1129500.0          Family    State servant   
2        1129500.0          Family    State servant   
3        1129500.0          Family    State servant   
4         135000.0   Unaccompanied          Working   

             NAME_EDUCATION_TYPE    NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  \
0  Secondary / secondary special  Single / not married  House / apartment   
1               Higher education               Married  House / apartment   
2               Higher education               Married  House / apartment   
3               Higher education               Married  House / apartment   
4  Secondary / secondary special  Single / not married  House / apartment   

   REGION_POPULATION_RELATIVE  DAYS_BIRTH  DAYS_EMPLOYED  DAYS_REGISTRATION  \
0                    0.018801       -9461           -637            -3648.0   
1                    0.003541      -16765          -1188            -1186.0   
2                    0.003541      -16765          -1188            -1186.0   
3                    0.003541      -16765          -1188            -1186.0   
4                    0.010032      -19046           -225            -4260.0   

   DAYS_ID_PUBLISH  OWN_CAR_AGE  FLAG_MOBIL  FLAG_EMP_PHONE  FLAG_WORK_PHONE  \
0            -2120          NaN           1               1                0   
1             -291          NaN           1               1                0   
2             -291          NaN           1               1                0   
3             -291          NaN           1               1                0   
4            -2531         26.0           1               1                1   

   FLAG_CONT_MOBILE  FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  CNT_FAM_MEMBERS  \
0                 1           1           0        Laborers              1.0   
1                 1           1           0      Core staff              2.0   
2                 1           1           0      Core staff              2.0   
3                 1           1           0      Core staff              2.0   
4                 1           1           0        Laborers              1.0   

   REGION_RATING_CLIENT  REGION_RATING_CLIENT_W_CITY  \
0                     2                            2   
1                     1                            1   
2                     1                            1   
3                     1                            1   
4                     2                            2   

  WEEKDAY_APPR_PROCESS_START  HOUR_APPR_PROCESS_START  \
0                  WEDNESDAY                       10   
1                     MONDAY                       11   
2                     MONDAY                       11   
3                     MONDAY                       11   
4                     MONDAY                        9   

   REG_REGION_NOT_LIVE_REGION  REG_REGION_NOT_WORK_REGION  \
0                           0                           0   
1                           0                           0   
2                        

In [ ]:
#save file
df.to_csv("merged_data_both.csv",header=True)

In [ ]:
!pip install pandas_profiling==3.0.0
!pip install pandas==1.3.1
from pandas_profiling import ProfileReport
df_check=df[["TARGET","NAME_CLIENT_TYPE"]]
profile = ProfileReport(df_check)
profile

# **Univariate Analysis**

In [ ]:
df_a.head()

In [ ]:
print("The percentage of people with no payment diffculty:")  
round((df_a["TARGET"].value_counts()[0]/len(df_a["TARGET"]))*100,2)

In [ ]:
print("Major Contract type:")
df_a["NAME_CONTRACT_TYPE"].value_counts()

In [ ]:
print("we can say that the ratio of female to male is 2:1")
df_a["CODE_GENDER"].value_counts()

In [ ]:
print("The people not owning a car to owing a car is 2:1")
df_a["FLAG_OWN_CAR"].value_counts()

In [ ]:
print("The percentage of people owing a house")
round(df_a["FLAG_OWN_REALTY"].value_counts()[0]/len(df_a),2)*100

In [ ]:
print("Distribution of the children\n",df_a["CNT_CHILDREN"].value_counts(),"\n\n")

plt.figure(figsize=(10,10))
c=(df_a["CNT_CHILDREN"].value_counts().values/len(df_a["CNT_CHILDREN"]))*100
l=list(df_a["CNT_CHILDREN"].value_counts().index)
l=list(map(str,l))
plt.bar(l,c,width=0.2)
plt.title("Seeing the percentage of children people are having")
plt.ylabel("percentage")
plt.xlabel("Number of children")
plt.show()



More people without children are taking loan.

The ratio is 70:30 

In [ ]:
print("Person with minmum salary")
print(df_a["AMT_INCOME_TOTAL"].min())
print("Person with maximum salary")
print(df_a["AMT_INCOME_TOTAL"].max())
print("The average of salary")
print(round(df_a["AMT_INCOME_TOTAL"].mean(),0))

In [ ]:
print("Maximum amount credited")
print(df_a["AMT_CREDIT"].max())

print("Average amount credited")
print(round(df_a["AMT_CREDIT"].mean(),0))


In [ ]:
print("Average cost of goods which are taken on credit")
print(round(df_a["AMT_GOODS_PRICE"].mean(),0))


In [ ]:
print("Family status of client")
df_a["NAME_TYPE_SUITE"].value_counts()

In [ ]:
print("Client's income type")
print(df_a["NAME_INCOME_TYPE"].value_counts())

In [ ]:
print("Client's education background")
print(df_a["NAME_EDUCATION_TYPE"].value_counts())
print("\n\n\n")
plt.bar(df_a["NAME_EDUCATION_TYPE"].value_counts().index,df_a["NAME_EDUCATION_TYPE"].value_counts().values/len(df_a["NAME_EDUCATION_TYPE"]),width=0.2)
plt.xticks(rotation=45)

In [ ]:
print("The day of applying loan")
print(df_a["WEEKDAY_APPR_PROCESS_START"].value_counts().sort_values())


Maximum client show up at tuesday and minimum at sunday

In [ ]:
print("The hour of day for applying loan")
print(df_a["HOUR_APPR_PROCESS_START"].value_counts())

In [ ]:
df_p.head(5)

In [ ]:
print("The status of client loan:")
ind=df_p["NAME_CONTRACT_STATUS"].value_counts().index
value=df_p["NAME_CONTRACT_STATUS"].value_counts()/len(df_p["NAME_CONTRACT_STATUS"])*100
[print(element) for element in list(zip(ind,value))]

In [ ]:
print("How the payment made to client:")
df_p["NAME_PAYMENT_TYPE"].value_counts()

In [ ]:
print("Rejection reason")
print(df_p["CODE_REJECT_REASON"].value_counts())

***We need to find xap value as it is important for not approving the loan***

In [ ]:
print("The client type")
df_p["NAME_CLIENT_TYPE"].value_counts()

we have more repeater client 

In [ ]:
print("client reqesting loan for\n\n")
df_p["NAME_GOODS_CATEGORY"].value_counts()
# clients are majorly applying loan for mobile,consumer electronics,computers

In [ ]:
df_a.head()

In [ ]:
threshold=10
cont_col=[]
disc_col=[]
for column in df_a.columns:
  if len(df_a[column].unique())>threshold:
    cont_col.append(column)
  else:
    disc_col.append(column)

print('*'*80)
print(f"continuous columns:\n{cont_col}")
print('*'*80)
print(f" discrete columns:\n {disc_col}")


In [ ]:
def insight_discrete(column):
  print(f"Percentage with respect to {column}")
  indexs=df_a[column].value_counts().index
  values=round(df_a[column].value_counts()/len(df_a[column]),2)
  [print(element) for element in list(zip(indexs,values))]
  print("\n\n")
  return 0

In [ ]:
for column in disc_col:
  insight_discrete(column)

In [ ]:
#knowing the distribution for continuous variables
def insight_continuous(column):
 
  plt.hist(df_a[column].dropna())
  plt.title(f"histogram for {column}")
  plt.xlabel(column)
  plt.ylabel("frequency")
  plt.show()
  print('\n\n')
  return 0

In [ ]:
for column in cont_col:
    insight_continuous(column)

In [ ]:
#doing for previous application features
threshold=10
cont_col_p=[]
disc_col_p=[]
for column in df_p.columns:
  if len(df_p[column].unique())>threshold:
    cont_col_p.append(column)
  else:
    disc_col_p.append(column)

print('*'*80)
print(f"continuous columns:\n{cont_col_p}")
print('*'*80)
print(f" discrete columns:\n {disc_col_p}")

In [ ]:
print("Removing unwanted column")
print(cont_col_p.pop(0))
print(cont_col_p.pop(0))


In [ ]:
def insight_discrete(column):
  print(f"Percentage with respect to {column}")
  indexs=df_p[column].value_counts().index
  values=round(df_p[column].value_counts()/len(df_p[column]),2)
  [print(element) for element in list(zip(indexs,values))]
  print("\n\n")
  return 0
for column in disc_col_p:
    insight_discrete(column)

In [ ]:
def insight_continuous(column):
 
  plt.hist(df_p[column].dropna())
  plt.title(f"histogram for {column}")
  plt.xlabel(column)
  plt.ylabel("frequency")
  plt.xticks(rotation=90)
  plt.show()
  print('\n\n')
  return 0

for column in disc_col_p:
  insight_continuous(column)

#bi variate EDA

**Columns in consideration** 
***************************************************

TARGET

NAME_CONTRACT_TYPE

CODE_GENDER

FLAG_OWN_CAR

FLAG_OWN_REALTY

CNT_CHILDREN

AMT_INCOME_TOTAL

NAME_TYPE_SUITE

NAME_INCOME_TYPE

NAME_EDUCATION_TYPE

NAME_FAMILY_STATUS

OCCUPATION_TYPE

CNT_FAM_MEMBERS

REGION_RATING_CLIENT

REGION_RATING_CLIENT_W_CITY

********************************************************

WEEKDAY_APPR_PROCESS_START

NAME_CASH_LOAN_PURPOSE

NAME_CONTRACT_STATUS

CODE_REJECT_REASON

NAME_CLIENT_TYPE

NAME_PORTFOLIO

NAME_CASH_LOAN_PURPOSE


In [ ]:
#plotting defaulter based on columns
plt.hist(df_a[df_a["TARGET"]==0]["NAME_CONTRACT_TYPE"],label="target=0",color='blue', alpha=0.6)
plt.hist(df_a[df_a["TARGET"]==1]["NAME_CONTRACT_TYPE"],label="target=1",color='red',alpha=0.6)
plt.xlabel('NAME_CONTRACT_TYPE')
plt.legend()
plt.show()

The number of cash loan clients are more.(frequency is more)



We have more defaulter in cash loans.

# **If we remove any nan values rows the data insight will be wrong. Always include the count i.e the distribution.Consider the base on which you are finding the percentage.**

In [ ]:
plt.hist(df_a[df_a["TARGET"]==0]["CODE_GENDER"],label="target=0",color='blue', alpha=0.6)
plt.hist(df_a[df_a["TARGET"]==1]["CODE_GENDER"],label="target=1",color='red',alpha=0.6)
plt.xlabel('CODE_GENDER')
plt.legend()
plt.show()

print("Defaluters")
print((df_a[df_a["TARGET"]==1]["CODE_GENDER"].value_counts()/len(df_a["CODE_GENDER"])*100))


The female clients are more.

The defaulters rate of female can be high due to high frequency.


Does having liability affect the target?

In [ ]:
plt.hist(df_a[df_a["TARGET"]==0]["FLAG_OWN_CAR"],label="target=0",color='blue', alpha=0.6)
plt.hist(df_a[df_a["TARGET"]==1]["FLAG_OWN_CAR"],label="target=1",color='red',alpha=0.6)
plt.xlabel('FLAG_OWN_CAR')
plt.legend()
plt.show()


It does not matter for having a liability

In [ ]:
plt.hist(df_a[df_a["TARGET"]==0]["FLAG_OWN_REALTY"],label="target=0",color='blue', alpha=0.6)
plt.hist(df_a[df_a["TARGET"]==1]["FLAG_OWN_REALTY"],label="target=1",color='red',alpha=0.6)
plt.xlabel('FLAG_OWN_REALTY')
plt.legend()
plt.show()


question on frequency high count high default

In [ ]:
plt.figure(figsize=(10,10))
plt.hist(df_a[df_a["TARGET"]==0]["CNT_CHILDREN"],label="target=0",color='blue', alpha=0.6, bins=100)
plt.hist(df_a[df_a["TARGET"]==1]["CNT_CHILDREN"],label="target=1",color='red',alpha=0.6, bins=100)
plt.xlabel('CNT_CHILDREN')
plt.legend()
plt.show()

In [ ]:

df_income=df_a[["AMT_INCOME_TOTAL","TARGET"]]
df_income['AMT_INCOME_ClASS'] = pd.qcut(df_income['AMT_INCOME_TOTAL'], q=10)
df_income.drop('AMT_INCOME_TOTAL',axis=1,inplace=True)
(df_income[df_income["TARGET"]==1]["AMT_INCOME_ClASS"].value_counts().sort_index()/len(df_income["TARGET"]))*100

Less the income more chance of becoming defaluter.

In [ ]:
df_a["NAME_TYPE_SUITE"]=df_a["NAME_TYPE_SUITE"].astype('str')
plt.figure(figsize=(10,10))
plt.hist(df_a[df_a["TARGET"]==0]["NAME_TYPE_SUITE"],label="target=0",color='blue', alpha=0.6,bins=50)
plt.hist(df_a[df_a["TARGET"]==1]["NAME_TYPE_SUITE"],label="target=1",color='red',alpha=0.6, bins=50)
plt.xlabel('NAME_TYPE_SUITE')
plt.legend()
plt.xticks(rotation=90)
plt.show()

In [ ]:
plt.hist(df_a[df_a["TARGET"]==0]["NAME_INCOME_TYPE"],label="target=0",color='blue', alpha=0.6,bins=50)
plt.hist(df_a[df_a["TARGET"]==1]["NAME_INCOME_TYPE"],label="target=1",color='red',alpha=0.6, bins=50)
plt.xlabel('NAME_INCOME_TYPE')
plt.legend()
plt.xticks(rotation=90)
plt.show()

In [ ]:
plt.hist(df_a[df_a["TARGET"]==0]["NAME_EDUCATION_TYPE"],label="target=0",color='blue', alpha=0.6,bins=50)
plt.hist(df_a[df_a["TARGET"]==1]["NAME_EDUCATION_TYPE"],label="target=1",color='red',alpha=0.6, bins=50)
plt.xlabel('NAME_EDUCATION_TYPE')
plt.legend()
plt.xticks(rotation=90)
plt.show()

In [ ]:
plt.hist(df_a[df_a["TARGET"]==0]["NAME_FAMILY_STATUS"],label="target=0",color='blue', alpha=0.6,bins=50)
plt.hist(df_a[df_a["TARGET"]==1]["NAME_FAMILY_STATUS"],label="target=1",color='red',alpha=0.6, bins=50)
plt.xlabel('NAME_FAMILY_STATUS')
plt.legend()
plt.xticks(rotation=90)
plt.show()

In [ ]:
plt.figure(figsize=(10,10))
df_a["OCCUPATION_TYPE"]=df_a["OCCUPATION_TYPE"].astype('str')
plt.hist(df_a[df_a["TARGET"]==0]["OCCUPATION_TYPE"],label="target=0",color='blue', alpha=0.6,bins=50)
plt.hist(df_a[df_a["TARGET"]==1]["OCCUPATION_TYPE"],label="target=1",color='red',alpha=0.6, bins=50)
plt.xlabel('OCCUPATION_TYPE')
plt.legend()
plt.xticks(rotation=90)
plt.show()

Nan value have to be known

In [ ]:
plt.figure(figsize=(10,10))
plt.hist(df_a[df_a["TARGET"]==0]["CNT_FAM_MEMBERS"],label="target=0",color='blue', alpha=0.6,bins=100)
plt.hist(df_a[df_a["TARGET"]==1]["CNT_FAM_MEMBERS"],label="target=1",color='red',alpha=0.6, bins=100)
plt.xlabel('CNT_FAM_MEMBERS')
plt.legend()
plt.xticks(rotation=90)
plt.show()

In [ ]:
#region
plt.hist(df_a[df_a["TARGET"]==0]["REGION_RATING_CLIENT"],label="target=0",color='blue', alpha=0.6,bins=100)
plt.hist(df_a[df_a["TARGET"]==1]["REGION_RATING_CLIENT"],label="target=1",color='red',alpha=0.6, bins=100)
plt.xlabel('REGION_RATING_CLIENT')
plt.legend()
plt.xticks(rotation=90)
plt.show()

In [ ]:
#city
plt.hist(df_a[df_a["TARGET"]==0]["REGION_RATING_CLIENT_W_CITY"],label="target=0",color='blue', alpha=0.6,bins=100)
plt.hist(df_a[df_a["TARGET"]==1]["REGION_RATING_CLIENT_W_CITY"],label="target=1",color='red',alpha=0.6, bins=100)
plt.xlabel('REGION_RATING_CLIENT_W_CITY')
plt.legend()
plt.xticks(rotation=90)
plt.show()

In [ ]:
#previous application day
plt.hist(df_a[df_a["TARGET"]==0]["WEEKDAY_APPR_PROCESS_START"],label="target=0",color='blue', alpha=0.6,bins=100)
plt.hist(df_a[df_a["TARGET"]==1]["WEEKDAY_APPR_PROCESS_START"],label="target=1",color='red',alpha=0.6, bins=100)
plt.xlabel('WEEKDAY_APPR_PROCESS_START')
plt.legend()
plt.xticks(rotation=90)
plt.show()

In [ ]:

plt.hist(df_p[df_p["NAME_CONTRACT_STATUS"]=="Approved"]["NAME_CASH_LOAN_PURPOSE"],label="Approved",color='blue', alpha=0.6,bins=100)
plt.hist(df_p[df_p["NAME_CONTRACT_STATUS"]=="Refused"]["NAME_CASH_LOAN_PURPOSE"],label="Refused",color='red',alpha=0.6, bins=100)
plt.xlabel('NAME_CASH_LOAN_PURPOSE')
plt.legend()
plt.xticks(rotation=90)
plt.show()

In [ ]:

plt.hist(df_p[df_p["NAME_CONTRACT_STATUS"]=="Approved"]["CODE_REJECT_REASON"],label="Approved",color='blue', alpha=0.6,bins=50)
plt.hist(df_p[df_p["NAME_CONTRACT_STATUS"]=="Refused"]["CODE_REJECT_REASON"],label="Refused",color='red',alpha=0.6, bins=50)
plt.xlabel('CODE_REJECT_REASON')
plt.legend()
plt.xticks(rotation=90)
plt.show()

In [ ]:
plt.hist(df_p[df_p["NAME_CONTRACT_STATUS"]=="Approved"]["NAME_CLIENT_TYPE"],label="Approved",color='blue', alpha=0.6,bins=50)
plt.hist(df_p[df_p["NAME_CONTRACT_STATUS"]=="Refused"]["NAME_CLIENT_TYPE"],label="Refused",color='red',alpha=0.6, bins=50)
plt.xlabel('NAME_CLIENT_TYPE')
plt.legend()
plt.xticks(rotation=90)
plt.show()

Repeater application is rejected at a higher rate

In [ ]:
plt.hist(df_p[df_p["NAME_CONTRACT_STATUS"]=="Approved"]["NAME_PORTFOLIO"],label="Approved",color='blue', alpha=0.6,bins=50)
plt.hist(df_p[df_p["NAME_CONTRACT_STATUS"]=="Refused"]["NAME_PORTFOLIO"],label="Refused",color='red',alpha=0.6, bins=50)
plt.xlabel('NAME_PORTFOLIO')
plt.legend()
plt.xticks(rotation=90)
plt.show()

client with previous application with cash have higher rate of refusing for allowance of loan.

# **Multi variate EDA**

keep the count for data and base of

NULL: Are male owning a car have greater default rate?


In [ ]:
(df_a[(df_a["TARGET"]==1) & (df_a["FLAG_OWN_CAR"]=="Y")]["CODE_GENDER"].value_counts()/len(df_a["TARGET"]))*100

Our null hypothesis is true.

Living with parent and  family status are linked with defaulter?

In [ ]:
(df_a[(df_a["TARGET"]==1) & (df_a["NAME_HOUSING_TYPE"]=='With parents')]["NAME_FAMILY_STATUS"].value_counts()/len(df_a["TARGET"]))*100

Is week day is related to approval?

In [ ]:
(df_p[df_p["NAME_CONTRACT_STATUS"]=='Approved']["WEEKDAY_APPR_PROCESS_START"].value_counts()/len(df_p["NAME_CONTRACT_STATUS"]))*100

In [ ]:
(df_p[df_p["NAME_CONTRACT_STATUS"]=='Refused']["WEEKDAY_APPR_PROCESS_START"].value_counts()/len(df_p["NAME_CONTRACT_STATUS"]))*100

NOT RELATED

Being a married women with less than or equal to two children are not a defaulter?

In [ ]:
(df_a[(df_a["TARGET"]==1) & (df_a["NAME_FAMILY_STATUS"]=='Married') & (df_a["CNT_CHILDREN"]<=2)]["CODE_GENDER"].value_counts()/len(df_a["TARGET"]))*100

In [ ]:
#seeing default=True based on gender with respect to  contract type
X_axis = np.arange(1,3)
per_genF_cont=round(df_a[(df_a["CODE_GENDER"]=="F") & (df_a["TARGET"]==1)]["NAME_CONTRACT_TYPE"].value_counts()/len(df_a["TARGET"])*100,2)
per_genM_cont=round(df_a[(df_a["CODE_GENDER"]=="M") & (df_a["TARGET"]==1)]["NAME_CONTRACT_TYPE"].value_counts()/len(df_a["TARGET"])*100,2)
plt.bar(x=X_axis-0.2, height=per_genF_cont,color="red",label="Female",width=0.2)
plt.bar(x=X_axis+0.2, height=per_genM_cont,color="blue",label="Male",width=0.2)
plt.xticks(X_axis,df_a["NAME_CONTRACT_TYPE"].unique())
plt.legend()
plt.show()  

In [ ]:
print("Tried to find out how gender and contract type are affecting target variable")
print("Gender=Female Target=1 contract_type")
print(per_genF_cont)
print("*"*80)
print("Gender=Male Target=1 contract_type")
print(per_genM_cont)

**Female are more in defalut rate when considered with contract type and male.**

In [ ]:

d={}
df_a[df_a["TARGET"]==1]["NAME_INCOME_TYPE"].value_counts()
ind_edu=df_a[df_a["TARGET"]==1]["NAME_EDUCATION_TYPE"].value_counts().index
for education in ind_edu:
  d[education]=round(df_a[(df_a["TARGET"]==1) & (df_a["NAME_EDUCATION_TYPE"]==education)]["NAME_INCOME_TYPE"].value_counts()/len(df_a["TARGET"]==1)*100,2)
  

Parameters in consideration : Target=1, income_source based on education.

In [ ]:
for key,value in d.items():
  print(f"Education:{key} \t\t \nDefalut rate based on income:\n{value}\n\n")

Client with Education background **"Secondary / secondary special"** with income type **"Working"**  has highest default that is **4.02%**		 


In [ ]:
d={}
d["1"]=df_a[(df_a["TARGET"]==1) & (df_a["REGION_RATING_CLIENT"]==1)]["NAME_CONTRACT_TYPE"].value_counts()/len(df_a["TARGET"])*100
d["2"]=df_a[(df_a["TARGET"]==1) & (df_a["REGION_RATING_CLIENT"]==2)]["NAME_CONTRACT_TYPE"].value_counts()/len(df_a["TARGET"])*100
d["3"]=df_a[(df_a["TARGET"]==1) & (df_a["REGION_RATING_CLIENT"]==3)]["NAME_CONTRACT_TYPE"].value_counts()/len(df_a["TARGET"])*100

In [ ]:
len(df_a["TARGET"])

In [ ]:
for key, value in d.items():
  print(f"REGION_RATING_CLIENT:{key}\n defaulters:\n{value} \n\n")

Overall percentage for defaulter in the dataset is of 8%.

**Region rating of 2 with cash loans has highest default i.e 5.43**

I think the confirmed parameters to train model should have:
1. Region rating
2. Gender
3. contract type
4. education
5. income source 



In [ ]:
print("Seeing the purpose of loan getting rejected:")
print(df_p[df_p["NAME_CONTRACT_STATUS"]=="Refused"]["NAME_CASH_LOAN_PURPOSE"].value_counts()/len(df_p["NAME_CONTRACT_STATUS"])*100)

**XNA,XAP are  major contributor for rejection in purpose of loan** 

In [ ]:
d={}
lst_client=df_p["NAME_CLIENT_TYPE"].unique()
for client in lst_client:
  d[client]=round(df_p[(df_p["NAME_CONTRACT_STATUS"]=="Refused") & (df_p["NAME_CLIENT_TYPE"]==client)]["NAME_CASH_LOAN_PURPOSE"].value_counts()/len(df_p["NAME_CONTRACT_STATUS"])*100,2)


In [ ]:
print("Seeing the rejection of contract based on the type of client based on client\n\n\n\n")
for key, value in d.items():
  print(f"client type: {key}\n purpose of loan with rejection percentage:\n{value}\n\n")

**We need to know the XAP and XNA Values in purpose of loan as they can been used for approving and rejection of loan.**

In [ ]:
d={}
d["1"]=df_a[(df_a["TARGET"]==1) & (df_a["REGION_RATING_CLIENT"]==1)]["OCCUPATION_TYPE"].value_counts()/len(df_a["TARGET"])*100
d["2"]=df_a[(df_a["TARGET"]==1) & (df_a["REGION_RATING_CLIENT"]==2)]["OCCUPATION_TYPE"].value_counts()/len(df_a["TARGET"])*100
d["3"]=df_a[(df_a["TARGET"]==1) & (df_a["REGION_RATING_CLIENT"]==3)]["OCCUPATION_TYPE"].value_counts()/len(df_a["TARGET"])*100


In [ ]:
for key, value in d.items():
  print(f"REGION_RATING_CLIENT:{key}\n defaulters based on occupation:\n{value} \n\n")

There is no such occupation related to client region rating with high value. The seems be distributed well.But "nan" value rates at top for all occupation need to find it out.